gen/gen_mpy.py

```diff
@@ -1968,6 +1968,17 @@ def try_generate_array_type(type_ast):
         replace('/','_div_')
     arr_to_c_convertor_name = 'mp_arr_to_%s' % array_convertor_suffix
     arr_to_mp_convertor_name = 'mp_arr_from_%s' % array_convertor_suffix
+
+    if hasattr(type_ast.type, 'names'):
```

Comment:
Is this really the best way to test for a struct type?  
There's already a function for it actually:

https://github.com/lvgl/lv_binding_micropython/blob/f960502a3260f795f21bfa90b2e7d7199430cf88/gen/gen_mpy.py#L287-L288

Reply:
the reason I am checking for the names attribute is because there are 2 different class instances that come through that function. One is IdentifierType which has the names attribute. the Other I cannot remember off the top of my head but it does not have the names attribute.  The type that comes through is not a c_ast.Struct or c_ast.Union instance It is an array and that array is filled with something. I am doing a check to make sure it is not filled with integers or bools as these cause problems when you set a point for each individual item in the array. We only want to use pointers for Structures and Unions that are in the array.

This is what we are looking for in terms of what `type_ast` should be
```
TypeDecl(declname='stops', quals=[], type=IdentifierType(names=['lv_gradient_stop_t' ))
```

and not this

```
TypeDecl(declname='inv_area_joined', quals=[], type=IdentifierType(names=['uint8_t'])
```

The only things that get processed by that function are
```
DIM: 2
TYPE: TypeDecl(declname='stops', quals=[], type=IdentifierType(names=['lv_gradient_stop_t']))

DIM: None
TYPE: TypeDecl(declname='points', quals=['const'], type=IdentifierType(names=['lv_point_t']))

DIM: 32
TYPE: TypeDecl(declname='inv_areas', quals=[], type=IdentifierType(names=['lv_area_t']))

DIM: 32
TYPE: TypeDecl(declname='inv_area_joined', quals=[], type=IdentifierType(names=['uint8_t']))

DIM: None
TYPE: TypeDecl(declname='col_dsc', quals=['const'], type=IdentifierType(names=['lv_coord_t']))

DIM: None
TYPE: TypeDecl(declname='props', quals=['const'], type=IdentifierType(names=['lv_style_prop_t']))
```

and of those only these ones actually end up having any changed made to them
```
DIM: 2
TYPE: TypeDecl(declname='stops', quals=[], type=IdentifierType(names=['lv_gradient_stop_t']))

DIM: 32
TYPE: TypeDecl(declname='inv_areas', quals=[], type=IdentifierType(names=['lv_area_t']))
```

that is because the others have a dim of None or the contents of the array is not a structure/union


and on the esp32 side of things
```
DIM: 6
TYPE: TypeDecl(declname='mac', quals=[], type=IdentifierType(names=['uint8_t']))

DIM: 4
TYPE: TypeDecl(declname='addr', quals=[], type=IdentifierType(names=['u32_t']))

DIM: 4
TYPE: TypeDecl(declname='tx_data', quals=[], type=IdentifierType(names=['uint8_t']))

DIM: 4
TYPE: TypeDecl(declname='client_ip', quals=[], type=IdentifierType(names=['u8_t']))

DIM: None
TYPE: TypeDecl(declname='nva', quals=['const'], type=Struct(name='sh2lib_nv', decls=None))

DIM: 10
TYPE: TypeDecl(declname='sta', quals=[], type=IdentifierType(names=['esp_netif_sta_info_t']))

DIM: None
TYPE: TypeDecl(declname='if_ip6', quals=[], type=IdentifierType(names=['esp_ip6_addr_t']))

DIM: None
TYPE: TypeDecl(declname='txt', quals=[], type=IdentifierType(names=['mdns_txt_item_t']))
```

this is what ends up getting changed
```
DIM: 10
TYPE: TypeDecl(declname='sta', quals=[], type=IdentifierType(names=['esp_netif_sta_info_t']))
```

Reply:
I am not sure wh `TypeDecl(declname='nva', quals=['const'], type=Struct(name='sh2lib_nv', decls=None))` is even making it to that function as it is a Structure and not an array. That is the reason why I am checking for the `names` attribute.

Reply:
There also seems to be an apparent problem somewhere because `inv_areas` does not have 32 dimensions and `stops` does not have 2 dimensions. They are both single dimension arrays and one has 32 elements in that array and the other has 2 elements in that array. a multidimensional array is a nested array as you know. in Python it would be 
```
some_list = [[0]] 
```

That is a 2 dimension array and there is a single element in the second dimensions. I do not believe that you can mix types and or dimensions in C code. meaning you cannot do this

```
some_list = ['hellow world', [0, 0, 0]]
```

so I am not really sure as to why pycparser has list of type names for the array

Reply:
as another note and it might be worth opening an issue for with pycparser. I think there should be a variable in `IdentifierType` that holds the "type" instance for what is actually being stored in the array instead of storing the name of the type. this way there would end up being a really solid way of being able to tell what the type is that is being stored in the array.Right now this kind of C code would much the whole thing up.

```
typedef uint8_t my_uint8_t;

my_uint8_t *my_array[15];
```

that is pseudo code and the syntax is probably wrong but you get the general idea.

What would end up happening is you would end up with pycparser doing this.

```
TypeDecl(declname='my_array', quals=[], type=IdentifierType(names=['my_uint8_t']))
```

and there is no concrete way of knowing if my_uint8_t is a struct, union or anything else. But if what was stored was the original declaration instead of just a name it would work out better.

That being said. in the gen_mpy.py script is there a name to pycparser type mapping?? if there is then maybe that can be used as the best way to nail down if it is a structure or a union.